### PR TITLE
[Feature] Tranding movies list, empty state, loader, search

### PR DESCRIPTION
**Pull request by Rammfall**

### CODE REVIEW

File: src/store/concepts/movies/operations/__mocks__/moviesResponse.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+      'adult': false,
+      'backdrop_path': '/2Fk3AB8E9dYIBc2ywJkxk8BTyhc.jpg',
+      'title': 'Greenland',
+      'genre_ids': [
+        28
+      ],
+      'poster_path': '/bNo2mcvSwIvnx8K6y1euAc1TLVq.jpg',
+      'original_language': 'en',
+      'popularity': 172.827,
+      'media_type': 'movie'
```
Comment (VladWithCookies) [inline src/store/concepts/movies/operations/__mocks__/moviesResponse.js:20]: `' '` for the keys are needless

File: src/store/concepts/movies/operations/__tests__/searchMovies.spec.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
+import { normalize, schema } from 'normalizr'
+
+import mockHttpClient from 'Api/__mocks__/mockHttpClient'
+import storeWithMiddlewareMock from 'Store/__mocks__/storeWithMiddlewareMock'
+import { API_REQUEST, API_SAVE, API_SUCCESS } from 'Store/concepts/data/types'
+
+import { moviesConstant, searchMovies } from '../../endpoints'
```
Comment (VladWithCookies) [inline src/store/concepts/movies/operations/__tests__/searchMovies.spec.js:7]: what is `moviesConstant`?

File: src/store/concepts/movies/operations/__tests__/searchMovies.spec.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+} from '../../types'
+
+import searchMoviesOperation from '../searchMovies'
+
+describe('searchMoviesOperation()', () => {
+  describe('with success response', () => {
+    const httpClient = mockHttpClient([
+      {
+        method: 'get',
+        resolve: { data: { ...trendingMoviesResponse } }
```
Comment (VladWithCookies) [inline src/store/concepts/movies/operations/__tests__/searchMovies.spec.js:22]: needless spread

File: src/store/concepts/movies/operations/searchMovies.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
+import { createLogic } from 'redux-logic'
+
+import { dataApiRequest, dataApiSave, dataApiSuccess } from 'Store/concepts/data/actions'
+import normalizeMovies from 'Store/schemas/movies'
+
```
Comment (VladWithCookies) [inline src/store/concepts/movies/operations/searchMovies.js:5]: why do you need this empty line?

File: src/store/concepts/movies/operations/searchMovies.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        total_results: totalResults
+      }
+    } = await httpClient.get(searchMovies, { params: { page, query } })
+
+    const normalizedResponse = normalizeMovies(results)
+
+    dispatch(dataApiSuccess({ endpoint: searchMovies }))
+    dispatch(dataApiSave({ endpoint: moviesConstant, response: normalizedResponse.entities.movies }))
+    dispatch(saveFoundedIds({ moviesIds: normalizedResponse.result }))
+    dispatch(saveFoundedQuantity({ quantity: totalResults }))
```
Comment (VladWithCookies) [inline src/store/concepts/movies/operations/searchMovies.js:37]: why do you need separate actions for this? why not to use the same actions as in `trendingMovies` operation?

File: src/store/concepts/movies/operations/trendingMovies.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        total_results: totalResults
+      }
+    } = await httpClient.get(trendingMovies, { params: { page } })
+
+    const normalizedResponse = normalizeMovies(results)
+
+    dispatch(dataApiSuccess({ endpoint: trendingMovies }))
+    dispatch(dataApiSave({ endpoint: moviesConstant, response: normalizedResponse.entities.movie }))
+    dispatch(saveTrendingMoviesIds({ moviesIds: normalizedResponse.result }))
+    dispatch(saveTrendingQuantity({ quantity: totalResults }))
```
Comment (VladWithCookies) [inline src/store/concepts/movies/operations/trendingMovies.js:36]: `quantity` -> `totalResults` or `total`

File: src/store/concepts/movies/reducer.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+const foundedQuantity = (state = null, action) => {
+  switch (action.type) {
+    case SAVE_FOUNDED_QUANTITY:
+      return action.quantity
+    default:
+      return state
+  }
+}
```
Comment (VladWithCookies) [inline src/store/concepts/movies/reducer.js:39]: you don't need these reducers

File: src/store/concepts/movies/selectors.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
+import { createSelector } from 'reselect'
+
+const dataSelector = (state) => state.data.movies
+
+export const getTrendingMoviesSelector = createSelector(
+  dataSelector,
+  ({ movies: { trendingMovies: { moviesIds } } }) => moviesIds,
```
Comment (VladWithCookies) [inline src/store/concepts/movies/selectors.js:7]: this will hurt memoization. create separate selector

File: src/store/schemas/movies.js [hunk 9]
```diff
@@ -0,0 +1,7 @@
+import { schema, normalize } from 'normalizr'
+
+const movie = new schema.Entity('movie')
+
```
Comment (VladWithCookies) [inline src/store/schemas/movies.js:4]: needless new line

File: src/utils/router/__tests__/changePage.spec.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+  describe('handle functions', () => {
+    changePage(history, handler)(2)
+
+    it('handles history.push', () => {
+      expect(history.push).toBeCalledTimes(1)
+    })
+
+    it('handles handler', () => {
+      expect(history.push).toBeCalledTimes(1)
```
Comment (VladWithCookies) [inline src/utils/router/__tests__/changePage.spec.js:21]: these tests are the same

File: src/utils/router/__tests__/changePage.spec.js [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+  const history = {
+    push: jest.fn()
+  }
+  const handler = jest.fn()
+
+  it('returns function', () => {
+    expect(typeof changePage(history, handler)).toStrictEqual('function')
+  })
+
+  describe('handle functions', () => {
```
Comment (VladWithCookies) [inline src/utils/router/__tests__/changePage.spec.js:13]: which function do you test?

File: src/utils/router/changePage.js [hunk 12]
```diff
@@ -0,0 +1,8 @@
+function changePage(history, handler) {
+  return (page) => {
+    history.push(`?page=${page}`)
+    handler(page)
+  }
+}
```
Comment (VladWithCookies) [inline src/utils/router/changePage.js:6]: need to be discussed. i don't understand the perpose of this one

File: src/utils/url/getSearchParams.js [hunk 13]
```diff
@@ -0,0 +1,9 @@
+/**
+ * @param {string} params
+ * @returns {URLSearchParams}
+ */
```
Comment (VladWithCookies) [inline src/utils/url/getSearchParams.js:4]: do you use some tool for annotation?

File: src/views/components/LoginForm/container.js [hunk 14]
```diff
@@ -3,9 +3,11 @@ import { withFormik } from 'formik'
 import * as yup from 'yup'
 
 import handleSubmit from 'Utils/formikEvents/handleSubmit'
-import LoginFormComponent from './component'
+
 import { defaultProps, propTypes } from './types'
```
Comment (VladWithCookies) [inline src/views/components/LoginForm/container.js:8]: so many new lines

File: src/views/components/MoviesList/__tests__/component.spec.js [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+
+  describe('with default props', () => {
+    it('matches snapshot', () => {
+      const component = shallow(<MoviesList {...defaultProps} />)
+
+      expect(component).toMatchSnapshot()
+    })
+  })
+
+  describe('when isLoading', () => {
```
Comment (VladWithCookies) [inline src/views/components/MoviesList/__tests__/component.spec.js:28]: `when isLoading is true`

File: src/views/components/MoviesList/component.js [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+
+import SearchForm from 'Views/components/SearchForm'
+import MovieCard from 'Views/components/shared/MovieCard'
+import PrivateLayout from 'Views/layouts/PrivateLayout'
+import Pagination from 'Views/components/Pagination'
+import MoviesLoader from 'Views/components/MoviesLoader'
+
+const MoviesList = ({ movies, quantity, isLoading, current, getMovies }) => (
+  <PrivateLayout>
+    <>
```
Comment (VladWithCookies) [inline src/views/components/MoviesList/component.js:13]: is fragment required?

File: src/views/components/MoviesList/component.js [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+      <SearchForm />
+      <div className='top-margin'>
+        {
+          isLoading
+            ? (
+              <MoviesLoader />
```
Comment (VladWithCookies) [inline src/views/components/MoviesList/component.js:19]: is this loader unique?

File: src/views/pages/Dashboard/container.js [hunk 18]
```diff
@@ -1,12 +1,69 @@
...
+  quantity: getTrendingMoviesQuantitySelector(state),
+  isLoading: loadingSelector(state, trendingMovies)
+})
+
+const mapDispatchToProps = {
+  getMovies: getTrendingMovies
+}
+
+export { MoviesList }
+export default connect(mapStateToProps, mapDispatchToProps)(withRouter(DashboardPage))
```
Comment (VladWithCookies) [inline src/views/pages/Dashboard/container.js:69]: use `compose` util

File: src/views/pages/Dashboard/container.js [hunk 19]
```diff
@@ -1,12 +1,69 @@
...
+  componentDidMount() {
+    const { getMovies } = this.props
+
+    getMovies(this.currentPage)
+  }
+
+  get currentPage() {
+    const { location: { search } } = this.props
+    const queryPage = getSearchParams(search).get('page')
+    return queryPage ? +queryPage : 1
```
Comment (VladWithCookies) [inline src/views/pages/Dashboard/container.js:25]: use `Number(someVariable)` for coercion to number

File: src/views/pages/Dashboard/__tests__/container.spec.js [hunk 20]
```diff
@@ -1,12 +1,70 @@
...
+          {...routerProps}
+        />
+      )
+      const container = diveTo(wrapper, MoviesList)
+
+      expect(container).toMatchSnapshot()
+    })
+  })
+
+  describe('when in url param page have value', () => {
```
Comment (VladWithCookies) [inline src/views/pages/Dashboard/__tests__/container.spec.js:51]: with page param

File: src/views/components/shared/MovieCard/component.js [hunk 21]
```diff
@@ -0,0 +1,31 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import { Card } from 'antd'
+
+import { imagePath } from 'Constants/index'
+
+const MovieCard = ({title, description, image}) => (
```
Comment (VladWithCookies) [inline src/views/components/shared/MovieCard/component.js:7]: unconsistent code style

File: src/views/components/shared/MovieCard/component.js [hunk 22]
```diff
@@ -0,0 +1,31 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import { Card } from 'antd'
+
+import { imagePath } from 'Constants/index'
```
Comment (VladWithCookies) [inline src/views/components/shared/MovieCard/component.js:5]: can we do `import { imagePath } from 'Constants';`

File: src/views/components/SearchForm/component.js [hunk 23]
```diff
@@ -0,0 +1,32 @@
...
+      md={{ span: 14 }}
+      lg={{ span: 12 }}
+      xl={{ span: 10 }}
+    >
+      <Input.Search
+        placeholder="Enter movie name"
+        size="large"
+        enterButton="Search"
+        className="top-margin"
+      />
```
Comment (VladWithCookies) [inline src/views/components/SearchForm/component.js:27]: do you plan to implement i18n for this?

File: src/views/components/Pagination/__tests__/container.spec.js [hunk 24]
```diff
@@ -0,0 +1,16 @@
...
+
+import Pagination from '../container'
+
+jest.mock('react-router-dom', () => ({
+  useHistory: jest.fn()
+}))
+
+describe('Pagination', () => {
+  it('matches snapshot', () => {
+    const container = shallow(<Pagination onChange={jest.fn()} />)
```
Comment (VladWithCookies) [inline src/views/components/Pagination/__tests__/container.spec.js:12]: `onChange={jest.fn}`

File: src/utils/url/getSearchParams.js (hunk 13 above)
Comment (Rammfall) [reply src/utils/url/getSearchParams.js:4]: It is a standard jsdoc, Enabled by default in WebStorm or with plugin for vscode

File: src/views/components/SearchForm/component.js (hunk 23 above)
Comment (Rammfall) [reply src/views/components/SearchForm/component.js:27]: Yes, after adding functionality for this component. In next PR, now search is not working

File: src/views/components/MoviesList/component.js (hunk 17 above)
Comment (Rammfall) [reply src/views/components/MoviesList/component.js:19]: Yes, only in movies page

File: src/views/components/MoviesList/component.js (hunk 16 above)
Comment (Rammfall) [reply src/views/components/MoviesList/component.js:13]: I fix prop type, Fragmet is useless

File: src/utils/url/getSearchParams.js (hunk 13 above)
Comment (VladWithCookies) [reply src/utils/url/getSearchParams.js:4]: do you plan to use it for each function in the project?

File: package.json [hunk 25]
```diff
@@ -78,6 +78,7 @@
     "lodash": "^4.17.20",
     "normalizr": "^3.6.1",
     "prop-types": "^15.7.2",
+    "ramda": "^0.27.1",
```
[the remark this answers is not in the document]
Comment (Rammfall) [reply package.json:81]: I remove lodash

File: src/assets/styles/app.scss [hunk 26]
```diff
@@ -61,3 +61,8 @@ body,
     display: none;
   }
 }
+
+.ant-row {
+  margin-left: 0 !important;
+  margin-right: 0 !important;
```
[the remark this answers is not in the document]
Comment (Rammfall) [reply src/assets/styles/app.scss:67]: I try

File: src/store/concepts/movies/__tests__/reducer.spec.js [hunk 27]
```diff
@@ -0,0 +1,57 @@
...
+
+describe('movies reducers', () => {
+  describe('trendingMovies reducers', () => {
+    describe('moviesIds reducer', () => {
+      it('handles SAVE_TRENDING_IDS', () => {
+        expect(movies(undefined, {
+          type: SAVE_TRENDING_IDS,
+          moviesIds: [1, 2, 3]
+        }).trendingMovies.moviesIds)
+          .toStrictEqual([
```
[the remark this answers is not in the document]
Comment (Rammfall) [reply src/store/concepts/movies/__tests__/reducer.spec.js:17]: Why? If we set https://www.npmjs.com/package/eslint-plugin-jest. Recommended use only toStrickEquual

File: src/utils/url/getSearchParams.js (hunk 13 above)
Comment (Rammfall) [reply src/utils/url/getSearchParams.js:4]: Yes!

**Pull request merged**